go/enclave/core/batchlimiter.go

```diff
@@ -0,0 +1,107 @@
+package core
+
+import (
+	"errors"
+
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/rlp"
+)
+
+// BatchSizeLimiter - Acts as a limiter for batches based
+// the data from the transaction that we have to publish to the l1.
+// Acts as a calldata reservation system that accounts for both
+// transactions and cross chain messages.
+type batchSizeLimiter struct {
+	Size            uint64
+	ContractAddress common.Address
```

Comment:
I would say that the logic to extract the cross chain messages is duplicated.
As they have to be included in the header as well.

What if you collect them at a higher lever, and then pass them into both this limiter and the header creator